kong/clustering/rpc/socket.lua

```diff
@@ -0,0 +1,287 @@
+-- socket represents an open WebSocket connection
+-- unlike the WebSocket object, it can be accessed via different requests
+-- with the help of semaphores
+
+
+local _M = {}
+local _MT = { __index = _M, }
+
+
+local cjson = require("cjson.safe")
+local future = require("kong.clustering.rpc.future")
+local utils = require("kong.clustering.rpc.utils")
+local queue = require("kong.clustering.rpc.queue")
+local jsonrpc = require("kong.clustering.rpc.json_rpc_v2")
+local constants = require("kong.constants")
+
+
+local assert = assert
+local cjson_encode = cjson.encode
+local cjson_decode = cjson.decode
+local is_timeout = utils.is_timeout
+local unpack = table.unpack
+local exiting = ngx.worker.exiting
+local ngx_time = ngx.time
+local ngx_log = ngx.log
+
+
+local CLUSTERING_PING_INTERVAL = constants.CLUSTERING_PING_INTERVAL
+local PING_WAIT = CLUSTERING_PING_INTERVAL * 1.5
+local PING_TYPE = "PING"
+local PONG_TYPE = "PONG"
+local ngx_WARN = ngx.WARN
+local ngx_DEBUG = ngx.DEBUG
+
+
+-- create a new socket wrapper, wb is the WebSocket object to use
+-- timeout and max_payload_len must already been set by caller when
+-- creating the `wb` object
+function _M.new(manager, wb, node_id)
+  local self = {
+    wb = wb,
+    interest = {}, -- id: callback pair
+    outgoing = queue.new(4096), -- write queue
+    manager = manager,
+    node_id = node_id,
+    sequence = 0,
+  }
+
+  return setmetatable(self, _MT)
+end
+
+
+function _M:_get_next_id()
+  local res = self.sequence
+  self.sequence = res + 1
```

Comment:
How about a random id? does ascending integer have some advantage?

Reply:
Random ID has no obvious benefit, and it creates chance for collision.

Since the `id` is only meaningful for the caller, using a sequential ID is the fastest and safest and simplest implementation and is the choice of most JSON-RPC clients.